File: backend/app/schemas/dataset.py
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


class DatasetColumnMapping(BaseModel):
    prompt: Optional[str] = None
    row_id: Optional[str] = None
    attack_category: Optional[str] = None
    severity: Optional[str] = None
    risk_level: Optional[str] = None
    subcategory: Optional[str] = None
    owasp_category: Optional[str] = None
    expected_safe_behavior: Optional[str] = None
    language: Optional[str] = None
    tags: Optional[str] = None


class DatasetImportRequest(BaseModel):
    source_type: Literal["direct_url", "github_raw", "huggingface", "kaggle"]
    source_uri: str = Field(..., min_length=5, max_length=2000)
    name: Optional[str] = None
    column_mapping: Optional[DatasetColumnMapping] = None

    # Request-only credentials. Do not store these in the database.
    kaggle_username: Optional[str] = None
    kaggle_key: Optional[str] = None
    kaggle_api_token: Optional[str] = None
    kaggle_file_path: Optional[str] = None
    hf_token: Optional[str] = None
    max_rows: int = Field(default=500, ge=1, le=500)


class DatasetCompatibilityReport(BaseModel):
    total_rows_detected: int
    valid_rows: int
    invalid_rows: int
    original_columns: list[str]
    detected_mapping: dict[str, Optional[str]]
    missing_prompt_rows: int
    detected_categories: list[str]
    detected_severities: list[str]
    owasp_mapping_coverage_percent: float
    ready_for_campaign: bool
    messages: list[str]


class DatasetRowResponse(BaseModel):
    id: int
    dataset_pk: int
    dataset_id: str
    row_id: str

    prompt: str
    attack_category: str
    severity: str
    risk_level: str
    subcategory: Optional[str]
    owasp_category: Optional[str]

    expected_safe_behavior: Optional[str]
    language: Optional[str]
    tags: list[str]
    row_metadata: dict[str, Any]

    created_at: datetime

    class Config:
        from_attributes = True


class DatasetResponse(BaseModel):
    id: int
    dataset_id: str

    name: str
    filename: str
    file_type: str

    source_type: str
    source_uri: Optional[str]

    row_count: int
    validation_status: str
    validation_message: Optional[str]

    original_columns: Optional[list[str]]
    detected_mapping: Optional[dict[str, Any]]
    validation_report: Optional[dict[str, Any]]

    created_at: datetime

    class Config:
        from_attributes = True


class DatasetListResponse(BaseModel):
    total: int
    items: list[DatasetResponse]


class DatasetUploadResponse(BaseModel):
    dataset: DatasetResponse
    preview_rows: list[DatasetRowResponse]


class DatasetImportResponse(BaseModel):
    dataset: DatasetResponse
    preview_rows: list[DatasetRowResponse]
    compatibility_report: DatasetCompatibilityReport


class DatasetRowsListResponse(BaseModel):
    dataset_id: str
    total: int
    items: list[DatasetRowResponse]

File: frontend/src/types/dataset.ts
export type DatasetSourceType =
  | "local_upload"
  | "direct_url"
  | "github_raw"
  | "huggingface"
  | "kaggle";

export type DatasetColumnMapping = {
  prompt?: string | null;
  row_id?: string | null;
  attack_category?: string | null;
  severity?: string | null;
  risk_level?: string | null;
  subcategory?: string | null;
  owasp_category?: string | null;
  expected_safe_behavior?: string | null;
  language?: string | null;
  tags?: string | null;
};

export type DatasetCompatibilityReport = {
  total_rows_detected: number;
  valid_rows: number;
  invalid_rows: number;
  original_columns: string[];
  detected_mapping: Record<string, string | null>;
  missing_prompt_rows: number;
  detected_categories: string[];
  detected_severities: string[];
  owasp_mapping_coverage_percent: number;
  ready_for_campaign: boolean;
  messages: string[];
};

export type DatasetRow = {
  id: number;
  dataset_pk: number;
  dataset_id: string;
  row_id: string;
  prompt: string;
  attack_category: string;
  severity: string;
  risk_level?: string;
  subcategory?: string | null;
  owasp_category?: string | null;
  expected_safe_behavior?: string | null;
  language?: string | null;
  tags: string[];
  row_metadata: Record<string, unknown>;
  created_at: string;
};

export type Dataset = {
  id: number;
  dataset_id: string;
  name: string;
  filename: string;
  file_type: string;
  source_type: string;
  source_uri?: string | null;
  row_count: number;
  validation_status: string;
  validation_message?: string | null;
  original_columns?: string[] | null;
  detected_mapping?: Record<string, unknown> | null;
  validation_report?: DatasetCompatibilityReport | null;
  created_at: string;
};

export type DatasetUploadResponse = {
  dataset: Dataset;
  preview_rows: DatasetRow[];
};

export type DatasetImportRequest = {
  source_type: "direct_url" | "github_raw" | "huggingface" | "kaggle";
  source_uri: string;
  name?: string;
  max_rows: number;
  column_mapping?: DatasetColumnMapping;

  // Request-only secrets. These are sent to backend for import only.
  hf_token?: string | null;
  kaggle_username?: string | null;
  kaggle_key?: string | null;
  kaggle_api_token?: string | null;
  kaggle_file_path?: string | null;
};

export type DatasetImportResponse = {
  dataset: Dataset;
  preview_rows: DatasetRow[];
  compatibility_report: DatasetCompatibilityReport;
};

export type DatasetListResponse = {
  total: number;
  items: Dataset[];
};

export type DatasetRowsListResponse = {
  dataset_id: string;
  total: number;
  items: DatasetRow[];
};
